### PR TITLE
Implement machine set status

### PR DESCRIPTION
**Pull request by k4leung4**
**What this PR does / why we need it**:
Implements the machineset status. Sets the replicas, fullylabeledreplicas, readyreplicas, availablereplicas fields.

Slight change to orphan adoptions, list all available machines, will not count machines with different owner refs, or non-matching labels.

@kubernetes/kube-deploy-reviewers

### CODE REVIEW

File: Gopkg.lock [hunk 1]
```diff
@@ -838,35 +838,64 @@
     "informers/storage/v1alpha1",
     "informers/storage/v1beta1",
     "kubernetes",
+    "kubernetes/fake",
```
Comment (jessicaochen) [inline Gopkg.lock]: I see changes to vendor metadata in the Gopkg.lock but no changes to the vendor directory. Is one of these not intentional? 

PS if you check in vendor changes, please do it in a different commit than your actual code for reviewer sanity.

File: pkg/controller/machineset/controller.go [hunk 2]
```diff
@@ -18,23 +18,55 @@ package machineset
...
+)
+
 // controllerKind contains the schema.GroupVersionKind for this controller type.
 var controllerKind = v1alpha1.SchemeGroupVersion.WithKind("MachineSet")
 
 // +controller:group=cluster,version=v1alpha1,kind=MachineSet,resource=machinesets
 type MachineSetControllerImpl struct {
 	builders.DefaultControllerFns
 
+	// kubernetesClient a client that knows hwo to consume Node resources
```
Comment (jessicaochen) [inline pkg/controller/machineset/controller.go]: spelling "how"

File: pkg/controller/machineset/controller.go [hunk 3]
```diff
@@ -18,23 +18,55 @@ package machineset
...
+	kubernetesClient kubernetes.Interface
+
 	// machineClient a client that knows how to consume Machine resources
 	machineClient machineclientset.Interface
```
Comment (jessicaochen) [inline pkg/controller/machineset/controller.go]: Looking at the way this is used, it seems to be more of a clusterapi interface since the code is pulling objects other than just machines from this.

File: pkg/controller/machineset/controller.go [hunk 4]
```diff
@@ -68,56 +102,38 @@ func (c *MachineSetControllerImpl) Reconcile(machineSet *v1alpha1.MachineSet) er
...
-			}
-		}
+	// Always updates status as machines come up or die.
+	updatedMS, err := updateMachineSetStatus(c.machineClient.ClusterV1alpha1().MachineSets(machineSet.Namespace), machineSet, newStatus)
+	if err != nil {
+		// Multiple things could lead to this update failing. Requeuing the replica set ensures
+		// Returning an error causes a requeue without forcing a hotloop
+		return err
+	}
+	// Resync the MachineSet after MinReadySeconds as a last line of defense to guard against clock-skew.
```
Comment (jessicaochen) [inline pkg/controller/machineset/controller.go]: It might just be me but why is clock skew involved with updating status?

File: pkg/controller/machineset/controller.go [hunk 5]
```diff
@@ -182,7 +219,266 @@ func (c *MachineSetControllerImpl) adoptOrphan(machineSet *v1alpha1.MachineSet,
...
+//
+// It groups the calls into batches, starting with a group of initialBatchSize.
+// Within each batch, it may call the function multiple times concurrently.
+//
+// If a whole batch succeeds, the next batch may get exponentially larger.
+// If there are any failures in a batch, all remaining batches are skipped
+// after waiting for the current batch to complete.
+//
+// It returns the number of successful calls to the function.
+func slowStartBatch(count int, initialBatchSize int, fn func() error) (int, error) {
```
Comment (jessicaochen) [inline pkg/controller/machineset/controller.go]: Can we break this out into a batch.go or something like that. This controller file is already getting kinda huge.

File: pkg/controller/machineset/controller.go (hunk 2 above)
Comment (k4leung4) [reply pkg/controller/machineset/controller.go]: done

File: pkg/controller/machineset/controller.go (hunk 3 above)
Comment (k4leung4) [reply pkg/controller/machineset/controller.go]: renamed to clusterAPIClient

File: pkg/controller/machineset/controller.go (hunk 4 above)
Comment (k4leung4) [reply pkg/controller/machineset/controller.go]: my understanding is that if there was a clock skew, and the machine may be seen as not available when it actually is due to minReadySeconds. without another event/reconcile, the status will not be updated to reflect that the machine is available. triggering a reconcile past the minReadySconds in the future, which force a recheck of the status will correctly update the availability status.

File: pkg/controller/machineset/controller.go (hunk 5 above)
Comment (k4leung4) [reply pkg/controller/machineset/controller.go]: removed this for the time being to avoid complexity
will investigate adding it back in a future change

File: pkg/controller/machineset/controller.go [hunk 6]
```diff
@@ -68,61 +87,43 @@ func (c *MachineSetControllerImpl) Reconcile(machineSet *v1alpha1.MachineSet) er
...
-				glog.Errorf("unable to delete a machine = %s, due to %v", machineToDelete.Name, err)
-				result = err
-			}
-		}
+	// Always updates status as machines come up or die.
+	updatedMS, err := updateMachineSetStatus(c.clusterAPIClient.ClusterV1alpha1().MachineSets(machineSet.Namespace), machineSet, newStatus)
+	if err != nil {
+		// Multiple things could lead to this update failing. Requeuing the replica set ensures
+		// Returning an error causes a requeue without forcing a hotloop
+		return err
```
Comment (jessicaochen) [inline pkg/controller/machineset/controller.go]: Would prefer this to wrapped in an error that explains the context. ie. machine set update failed.

File: pkg/controller/machineset/controller.go [hunk 7]
```diff
@@ -68,61 +87,43 @@ func (c *MachineSetControllerImpl) Reconcile(machineSet *v1alpha1.MachineSet) er
...
 	}
+	// Resync the MachineSet after MinReadySeconds as a last line of defense to guard against clock-skew.
+	if manageReplicasErr == nil && updatedMS.Spec.MinReadySeconds > 0 &&
+		updatedMS.Status.ReadyReplicas == *(updatedMS.Spec.Replicas) &&
```
Comment (jessicaochen) [inline pkg/controller/machineset/controller.go]: Need more commentary to understand this scenario. I do not understand the scenario where all replicas are ready but some are not available. Shouldn't a replica be available to be counted as ready?

File: pkg/controller/machineset/controller.go [hunk 8]
```diff
@@ -182,7 +204,218 @@ func (c *MachineSetControllerImpl) adoptOrphan(machineSet *v1alpha1.MachineSet,
...
+
+	// Save the generation number we acted on, otherwise we might wrongfully indicate
+	// that we've seen a spec update when we retry.
+	// TODO: This can clobber an update if we allow multiple agents to write to the
+	// same status.
+	newStatus.ObservedGeneration = ms.Generation
+
+	var getErr, updateErr error
+	var updatedMS *v1alpha1.MachineSet
+	for i, ms := 0, ms; ; i++ {
```
Comment (jessicaochen) [inline pkg/controller/machineset/controller.go]: This looks interesting. What does the additional ms:=0, ms do?

File: pkg/controller/machineset/controller.go [hunk 9]
```diff
@@ -182,7 +204,218 @@ func (c *MachineSetControllerImpl) adoptOrphan(machineSet *v1alpha1.MachineSet,
...
+	for _, machine := range filteredMachines {
+		if templateLabel.Matches(labels.Set(machine.Labels)) {
+			fullyLabeledReplicasCount++
+		}
+		node, err := c.getMachineNode(machine)
+		if err != nil {
+			glog.Warningf("Unable to get node for machine %v, %v", machine.Name, err)
+			continue
+		}
+		if isNodeReady(node) {
```
Comment (jessicaochen) [inline pkg/controller/machineset/controller.go]: I would have assumed that a ready node is also available. Is that k8s naming convention or something you can choose to be more intuitive?

Comment (k4leung4) [reply pkg/controller/machineset/controller.go]: This is the mapping of the machineset status for replicas. 
A node can be ready, but not available. Whether it is available or not is dependent on what the minReadySeconds setting is.
A node will not be ready if minReadySeconds have not elapsed since last transition.

File: pkg/controller/machineset/controller.go (hunk 8 above)
Comment (k4leung4) [reply pkg/controller/machineset/controller.go]: this was copied from replicaset controller, which now that you point out, isnt really needed.
after checking with krousey, this sets i to 0, and ms to ms, multi-variable initialization.
but since ms is already a variable, it doesnt do anything.
fixed it to be a simple i while loop.

File: pkg/controller/machineset/controller.go (hunk 7 above)
Comment (k4leung4) [reply pkg/controller/machineset/controller.go]: added more comments.

	// Clock-skew is an issue as it may impact whether a ready replica is counted as an available replica.
	// A replica is available if the amount of time since last transition exceeds MinReadySeconds.
	// If there was a clock skew, checking whether the amount of time since last transition to ready state
	// exceeds MinReadySeconds could be incorrect.
	// To avoid an available replica stuck in the ready state, we force a reconcile after MinReadySeconds,
	// at which point it should confirm any available replica to be available.

**Comment by k4leung4** [comment]
This PR is no longer needed due to the landing of PR #180 and #193 

/close
